File: js/waves.js
'use strict';

function shuffle(a) {
  for (var i = a.length - 1; i > 0; i--) {
    var j = Math.floor(Math.random() * (i + 1));
    var tmp = a[i]; a[i] = a[j]; a[j] = tmp;
  }
  return a;
}

var WAVE = {
  BOSSES: ['dragon', 'orcKing', 'lord', 'iceDragon', 'warMachine', 'voidLord'],

  // sesgo de composición según el tema del mapa
  THEME_FLAVOR: {
    plains: { n: 12, picks: [['skeleton', 0.5], ['orc', 0.4], ['undead', 0.4]] },
    desert: { n: 11, picks: [['skeleton', 0.7], ['undead', 0.5], ['fireGolem', 0.4]] },
    forest: { n: 11, picks: [['treant', 0.7], ['wisp', 0.6], ['troll', 0.4], ['crawler', 0.5]] },
    frozen: { n: 10, picks: [['iceWraith', 0.7], ['stoneGolem', 0.5], ['stormSpirit', 0.5]] },
    void: { n: 9, picks: [['voidWalker', 0.8], ['hulker', 0.5], ['crawler', 0.5], ['demon', 0.5]] }
  },

  build: function (n) {
    return this.buildFor(n, 'plains');
  },

  buildFor: function (n, mapId) {
    if (n === 0) return [];
    var map = (typeof MAPS_BY_ID !== 'undefined' && MAPS_BY_ID[mapId]) || MAPS_BY_ID['plains'];
    var theme = map ? map.theme : 'plains';
    var list = [];
    var gap = Math.max(0.35, 1.4 - n * 0.06);

    if (n % 5 === 0) {
      var idx = ((n / 5) - 1) % WAVE.BOSSES.length;
      var boss = WAVE.BOSSES[idx];
      var esc = [];
      var bg = Math.max(0.3, 1.1 - n * 0.05);
      // Escoltas moderadas: el jefe debe ser la amenaza principal, no el muro
      var goblins = Math.floor(n * 1.15);
      for (var i = 0; i < goblins; i++) esc.push({ type: 'goblin', gap: bg });
      var orcs = Math.floor(n * 0.45);
      for (var j = 0; j < orcs; j++) esc.push({ type: 'orc', gap: bg });
      if (n >= 10) {
        var sks = Math.floor(n * 0.3);
        for (var k = 0; k < sks; k++) esc.push({ type: 'skeleton', gap: bg });
      }
      if (n >= 15) {
        var uds = Math.floor(n * 0.25);
        for (var m = 0; m < uds; m++) esc.push({ type: 'undead', gap: bg });
      }
      if (n >= 20 && boss !== 'iceDragon') {
        for (var sp = 0; sp < Math.floor(n * 0.25); sp++) esc.push({ type: 'stormSpirit', gap: bg });
      }
      // séquito especializado: sanadores y escudos protegen al jefe
      if (n >= 15) {
        var mend = Math.min(3, 1 + Math.floor(n / 25));
        for (var md = 0; md < mend; md++) esc.push({ type: 'mender', gap: bg });
        var shd = Math.floor(n / 12);
        for (var os2 = 0; os2 < shd; os2++) esc.push({ type: 'orcShield', gap: bg });
      }
      esc = shuffle(esc);
      var bossHpMul = 1 + (n - 5) * 0.08;
      esc.unshift({ type: boss, gap: 3, bossHpMul: bossHpMul });
      list = esc;
    } else {
      var types = [];
      types.push(['goblin', 3 + n * 2]);
      if (n >= 3) types.push(['orc', Math.floor(n * 1.2)]);
      if (n >= 5) types.push(['skeleton', Math.floor(n * 0.9)]);
      if (n >= 6) types.push(['crawler', Math.floor((n - 4) * 0.9)]);
      if (n >= 7) types.push(['berserker', Math.floor((n - 4) * 0.9)]);
      if (n >= 8) types.push(['gargoyle', Math.floor((n - 6) * 0.5)]);
      if (n >= 9) types.push(['bat', Math.floor(n * 0.6)]);
      if (n >= 11) types.push(['sorcerer', Math.floor((n - 8) * 0.5)]);
      if (n >= 11) types.push(['wisp', Math.floor((n - 9) * 0.5)]);
      if (n >= 12) types.push(['shaman', Math.floor((n - 10) * 0.35)]);
      if (n >= 13) types.push(['troll', Math.floor((n - 10) * 0.5)]);
      if (n >= 14) types.push(['fireGolem', Math.floor((n - 12) * 0.4)]);
      if (n >= 14) types.push(['demon', Math.floor((n - 12) * 0.45)]);
      if (n >= 15) types.push(['iceWraith', Math.floor((n - 12) * 0.5)]);
      if (n >= 16) types.push(['undead', Math.floor((n - 12) * 0.5)]);
      if (n >= 17) types.push(['stoneGolem', Math.floor((n - 14) * 0.4)]);
      if (n >= 17) types.push(['lich', Math.floor((n - 15) * 0.35)]);
      if (n >= 18) types.push(['treant', Math.floor((n - 15) * 0.5)]);
      if (n >= 18) types.push(['necromancer', Math.floor((n - 14) * 0.4)]);
      if (n >= 19) types.push(['stormSpirit', Math.floor((n - 15) * 0.5)]);
      if (n >= 19) types.push(['voidWalker', Math.floor((n - 17) * 0.5)]);
      if (n >= 13) types.push(['orcShield', Math.floor((n - 11) * 0.45)]);
      if (n >= 15) types.push(['mender', Math.floor((n - 13) * 0.35)]);
      if (n >= 17) types.push(['phaseStalker', Math.floor((n - 15) * 0.45)]);
      if (n >= 8) types.push(['saboteur', Math.floor((n - 6) * 0.4)]);
      if (n >= 10) types.push(['assassin', Math.floor((n - 8) * 0.45)]);
      if (n >= 12) types.push(['thief', Math.floor((n - 10) * 0.3)]);
      if (n >= 18) types.push(['hulker', Math.floor((n - 16) * 0.3)]);
      if (n >= 7) types.push(['splitter', Math.floor((n - 5) * 0.4)]);
      for (var t = 0; t < types.length; t++) {
        var arr = types[t];
        for (var c = 0; c < arr[1]; c++) list.push({ type: arr[0], gap: gap });
      }
      // sabor del tema
      var fl = WAVE.THEME_FLAVOR[theme];
      if (fl && n >= fl.n) {
        for (var f = 0; f < fl.picks.length; f++) {
          var pk = fl.picks[f];
          var count = Math.floor((n - fl.n + 3) * pk[1]);
          for (var f2 = 0; f2 < count; f2++) list.push({ type: pk[0], gap: gap });
        }
      }
      list = shuffle(list);
      if (list.length) list[0].gap = 2;
    }
    if (list.length > 90) list = list.slice(0, 90);
    return list;
  }
};

File: python/waves.py
import random
import math


def shuffle(a):
    for i in range(len(a) - 1, 0, -1):
        j = int(random.random() * (i + 1))
        a[i], a[j] = a[j], a[i]
    return a


class Wave:
    BOSSES = ['dragon', 'orcKing', 'lord', 'iceDragon', 'warMachine', 'voidLord']

    THEME_FLAVOR = {
        'plains': {'n': 12, 'picks': [['skeleton', 0.5], ['orc', 0.4], ['undead', 0.4]]},
        'desert': {'n': 11, 'picks': [['skeleton', 0.7], ['undead', 0.5], ['fireGolem', 0.4]]},
        'forest': {'n': 11, 'picks': [['treant', 0.7], ['wisp', 0.6], ['troll', 0.4], ['crawler', 0.5]]},
        'frozen': {'n': 10, 'picks': [['iceWraith', 0.7], ['stoneGolem', 0.5], ['stormSpirit', 0.5]]},
        'void': {'n': 9, 'picks': [['voidWalker', 0.8], ['hulker', 0.5], ['crawler', 0.5], ['demon', 0.5]]},
    }

    def build(self, n):
        return self.build_for(n, 'plains')

    def build_for(self, n, map_id):
        if n == 0:
            return []
        from maps import MAPS_BY_ID
        m = MAPS_BY_ID.get(map_id, MAPS_BY_ID.get('plains'))
        theme = m['theme'] if m else 'plains'
        lst = []
        gap = max(0.35, 1.4 - n * 0.06)

        if n % 5 == 0:
            idx = ((n // 5) - 1) % len(self.BOSSES)
            boss = self.BOSSES[idx]
            esc = []
            bg = max(0.3, 1.1 - n * 0.05)
            goblins = int(n * 1.5)
            for _ in range(goblins):
                esc.append({'type': 'goblin', 'gap': bg})
            orcs = int(n * 0.6)
            for _ in range(orcs):
                esc.append({'type': 'orc', 'gap': bg})
            if n >= 10:
                sks = int(n * 0.4)
                for _ in range(sks):
                    esc.append({'type': 'skeleton', 'gap': bg})
            if n >= 15:
                uds = int(n * 0.3)
                for _ in range(uds):
                    esc.append({'type': 'undead', 'gap': bg})
            if n >= 20 and boss != 'iceDragon':
                for _ in range(int(n * 0.3)):
                    esc.append({'type': 'stormSpirit', 'gap': bg})
            shuffle(esc)
            boss_hp_mul = 1 + (n - 5) * 0.08
            esc.insert(0, {'type': boss, 'gap': 3, 'bossHpMul': boss_hp_mul})
            lst = esc
        else:
            types = []
            types.append(['goblin', 3 + n * 2])
            if n >= 3:
                types.append(['orc', int(n * 1.2)])
            if n >= 5:
                types.append(['skeleton', int(n * 0.9)])
            if n >= 6:
                types.append(['crawler', int((n - 4) * 0.9)])
            if n >= 7:
                types.append(['berserker', int((n - 4) * 0.9)])
            if n >= 8:
                types.append(['gargoyle', int((n - 6) * 0.5)])
            if n >= 9:
                types.append(['bat', int(n * 0.6)])
            if n >= 11:
                types.append(['sorcerer', int((n - 8) * 0.5)])
                types.append(['wisp', int((n - 9) * 0.5)])
            if n >= 12:
                types.append(['shaman', int((n - 10) * 0.35)])
            if n >= 13:
                types.append(['troll', int((n - 10) * 0.5)])
            if n >= 14:
                types.append(['fireGolem', int((n - 12) * 0.4)])
                types.append(['demon', int((n - 12) * 0.45)])
            if n >= 15:
                types.append(['iceWraith', int((n - 12) * 0.5)])
            if n >= 16:
                types.append(['undead', int((n - 12) * 0.5)])
            if n >= 17:
                types.append(['stoneGolem', int((n - 14) * 0.4)])
                types.append(['lich', int((n - 15) * 0.35)])
            if n >= 18:
                types.append(['treant', int((n - 15) * 0.5)])
                types.append(['necromancer', int((n - 14) * 0.4)])
            if n >= 19:
                types.append(['stormSpirit', int((n - 15) * 0.5)])
                types.append(['voidWalker', int((n - 17) * 0.5)])
            if n >= 8:
                types.append(['saboteur', int((n - 6) * 0.4)])
            if n >= 10:
                types.append(['assassin', int((n - 8) * 0.45)])
            if n >= 12:
                types.append(['thief', int((n - 10) * 0.3)])
            if n >= 18:
                types.append(['hulker', int((n - 16) * 0.3)])
            if n >= 7:
                types.append(['splitter', int((n - 5) * 0.4)])
            for t_type, count in types:
                for _ in range(count):
                    lst.append({'type': t_type, 'gap': gap})
            fl = self.THEME_FLAVOR.get(theme)
            if fl and n >= fl['n']:
                for pk_type, pk_weight in fl['picks']:
                    count = int((n - fl['n'] + 3) * pk_weight)
                    for _ in range(count):
                        lst.append({'type': pk_type, 'gap': gap})
            shuffle(lst)
            if lst:
                lst[0]['gap'] = 2

        if len(lst) > 90:
            lst = lst[:90]
        return lst


WAVE = Wave()
